Implement UI for Admin Search of Hashtags

ThisIsMissEm opened the pull request:
This is a very rough first pass, but gives the following UI:

<img width="1243" src="https://github.com/mastodon/mastodon/assets/30827/0ced7bff-2532-44f9-9b27-e033c0e00c6f">

And with more than 20 tags:

![image](https://github.com/mastodon/mastodon/assets/30827/38a60474-e4d6-4958-a926-5e7ddd1bb18b)


Implements #29986, without the dynamic autocomplete for now.

ThisIsMissEm commented:
@renchap do we have any tooling in the admin UI to upgrade a text input to a searchable text input? I don't really want to make this a react widget by default, but instead enhance it in some way?

ThisIsMissEm commented on app/views/admin/tags/index.html.haml:
```diff
@@ -0,0 +1,39 @@
+    - if @tags.empty?
+      = nothing_here 'nothing-here--under-tabs'
+    - else
+      = render partial: 'admin/trends/tags/tag', collection: @tags, locals: { f: nil }
```
Arguably maybe this partial should be moved here (`admin/tags`) and referenced from `admin/trends/tags`

The `f: nil` thing is simply to remove the checkboxes, since in this view we're not doing bulk actions right now, and tbh, I don't know if bulk actions would be desired here.

ThisIsMissEm commented on app/models/tag.rb:
```diff
@@ -133,7 +133,8 @@ def find_or_create_by_names(name_or_names)
       stripped_term = term.strip
 
-      query = Tag.listable.matches_name(stripped_term)
+      query = Tag.matches_name(stripped_term)
+      query = query.merge(Tag.listable) unless options[:include_unlistable]
```
This change was necessary because in the Admin UI we want to see all hashtags when we search for a hashtag, so admin sets `include_unlistable` to true; I thought this was the cleaner way of handling this, instead of an `exclude_unlistable` and then trying to find all the code paths that used `search_for` to make them add that option.

ClearlyClaire replied:
Having something like `{ exclude_unlistable: true }.merge(options)` would allow a `exclude_unlistable` option without changing the call sites, but I think `include_unlistable` is fine too.

ThisIsMissEm replied:
Oooh, that could be a good call. I think I like that better given there's already an exclude_... for something else here

oneiros replied:
Do you want to make that change then? Personally I believe both approaches are perfectly fine, but I agree it could make for a nicer "symmetry" with `exclude_unreviewed`.

ThisIsMissEm replied:
Yeah, I will, have just been dealing with other things this week

ThisIsMissEm commented on spec/models/admin/tag_filter_spec.rb:
```diff
@@ -0,0 +1,16 @@
+    it 'returns filtered tags' do
+      filter = described_class.new(name: 'test')
+
+      expect(filter.results).to eq([listable_tag, not_listable_tag])
```
This test should exercise the code change in `app/models/tag.rb` to allow searching outside of listable tags.

ThisIsMissEm commented on spec/models/tag_spec.rb:
```diff
@@ -232,5 +232,23 @@
+      tag = Fabricate(:tag, name: 'match')
+      unlisted_tag = Fabricate(:tag, name: 'matchunlisted', listable: false)
+
+      results = described_class.search_for('match', 5, 0, include_unlistable: true)
```
Arguably the limit / offset here should probably be via an option as well, but that involves modifying code well out of scope of this PR, hence leaving as is for now.

ThisIsMissEm commented on config/locales/en.yml:
```diff
@@ -880,7 +880,23 @@ en:
+      reset: Reset
       review: Review status
+      search: Search
+      title: Tags
```
```suggestion
      title: Hashtags
```

ThisIsMissEm commented on config/locales/en.yml:
```diff
@@ -880,7 +880,23 @@ en:
         action: Check here for more information
         message_html: "<strong>Your object storage is misconfigured. The privacy of your users is at risk.</strong>"
     tags:
+      back_to_tags: Back to tags
```
```suggestion
      back_to_tags: Back to hashtags
```

shleeable commented:
Request: Please add the ability to rename the case sensitivity if that's possible.

`#youtube` to `#YouTube` etc etc.

ClearlyClaire commented:
Looks ok overall but I'd need to have a more careful review.

> Request: Please add the ability to rename the case sensitivity if that's possible.
> 
> `#youtube` to `#YouTube` etc etc.

I'm not sure what you mean, isn't that already possible?

ThisIsMissEm commented:
> Request: Please add the ability to rename the case sensitivity if that's possible.
> 
> `#youtube` to `#YouTube` etc etc.

That's out of scope for this PR. But if there's an issue, maybe it'd be possible for me to work on it, but we'd have to work out how the rename would affect already published posts.

oneiros commented on app/models/admin/tag_filter.rb:
```diff
@@ -0,0 +1,74 @@
+    scope = Tag.reorder(nil)
+
+    params.each do |key, value|
+      next if key.to_s == 'page'
```
Why the `#to_s`? You did call `#symbolize_keys` on the params hash in the constructor, so keys should already be symbols. Or am I missing something?

ThisIsMissEm replied:
This was based largely ob the AccountFilter, and I don't know enough about kaminari and our pagination helpers to say with confidence that a change would be okay

oneiros replied:
> This was based largely ob the AccountFilter

Ah, I see. Well, there is the same oddity there. But I see no need to repeat it here :slightly_smiling_face: 

> and I don't know enough about kaminari and our pagination helpers to say with confidence that a change would be okay

I do not believe pagination would be affected here. Pagination happens in the controller and uses the original params, not the ones from this class (that were converted to a hash with symbols as keys).

oneiros commented on app/models/admin/tag_filter.rb:
```diff
@@ -0,0 +1,74 @@
+  private
+
+  def scope_for(key, value)
+    case key.to_s
```
Same as above: Keys should be symbols, why do the extra step of converting them?

oneiros commented on app/views/admin/trends/tags/_tag.html.haml:
```diff
@@ -1,6 +1,7 @@
 .batch-table__row{ class: [tag.requires_review? && 'batch-table__row--attention', !tag.requires_review? && !tag.trendable? && 'batch-table__row--muted'] }
-  %label.batch-table__row__select.batch-table__row__select--aligned.batch-checkbox
-    = f.check_box :tag_ids, { multiple: true, include_hidden: false }, tag.id
+  - unless f.nil?
```
An alternative could be to use

```haml
- if local_assigns[:f]
```

I always found this to be quite ugly, but it would mean you could get rid of the `f: nil` argument when rendering this partial from `app/views/admin/tags/index.html.haml`.

ThisIsMissEm replied:
This seems like it could be nice, but I'm thinking I actually want to refactor this to not use that partial since I'm not sure how useful it is

ThisIsMissEm replied:
Have gone with a custom tag row partial, which emphasizes the moderation status as a whole, instead of emphasizing the trending status.

oneiros commented:
I left a couple of small remarks, but overall I think this is fantastic!

As for the suggestions to rename "tag" to "hashtag" in the locales, I agree with this change. I looked around the UI a bit and "hashtag" seems to be the wording commonly used. So for consistency's sake, I would agree to making these changes.

Lastly, as codecov is complaining, I believe it would be great to have some more examples in `spec/models/admin/tag_filter_spec.rb`.

ThisIsMissEm commented:
> Lastly, as codecov is complaining, I believe it would be great to have some more examples in spec/models/admin/tag_filter_spec.rb.

@oneiros yeah, I wasn't sure whether the filters were actually useful, since you're always searching for a tag pretty much.

So more looking for user feedback here: do tje filters make sense? Do we need them? Should we change them?

ThisIsMissEm committed:
Clarify that unusable hashtags are only usable locally, and do not affect remote statuses

ThisIsMissEm commented:
Here's the updated screenshots:

<img width="1133" alt="Screenshot 2024-07-05 at 23 55 46" src="https://github.com/mastodon/mastodon/assets/30827/4b57bbe0-783b-4dbd-924b-36958be14d03">
<img width="1187" alt="Screenshot 2024-07-05 at 23 55 52" src="https://github.com/mastodon/mastodon/assets/30827/dc80af2c-4d49-4153-a2af-ee8dbdbafb95">

I've also dropped the "order by pending" since it didn't really make sense in practice. 

The trending hashtags UI uses an icon for the hash, I'm not sure why, but it looked rather weird with the new tag row partial, so I switched to just using a textual # instead.

ThisIsMissEm commented on config/locales/simple_form.en.yml:
```diff
@@ -311,7 +311,7 @@ en:
         name: Hashtag
         trendable: Allow this hashtag to appear under trends
-        usable: Allow posts to use this hashtag
+        usable: Allow posts to use this hashtag locally
```
See: https://github.com/mastodon/mastodon/blob/main/app/validators/disallowed_hashtags_validator.rb

ThisIsMissEm commented on app/javascript/styles/mastodon/accounts.scss:
```diff
@@ -346,6 +346,10 @@
       font-weight: 700;
     }
+
+    .warning-hint {
```
I'm not 100% sold on this, but given all hashtags are unreviewed by default, the "pending review" became very overbearing in the list view with font-weight 500.

ThisIsMissEm commented on app/models/tag.rb:
```diff
@@ -132,8 +133,10 @@ def find_or_create_by_names(name_or_names)
 
     def search_for(term, limit = 5, offset = 0, options = {})
       stripped_term = term.strip
+      options = { exclude_unlistable: true, exclude_unreviewed: false }.merge(options)
```
I could've used `merge!` here, but that felt kind of wrong for some reason

oneiros replied:
That is one reason `ActiveSupport` adds `#reverse_merge`. So you could do:

```ruby
options.reverse_merge!({ exclude_unlistable: true, exclude_unreviewed: false })
```

(I just wanted to provide an alternative, but do not have a strong preference myself.)

ThisIsMissEm commented on app/views/admin/tags/_tag.html.haml:
```diff
@@ -0,0 +1,28 @@
+.batch-table__row{ class: [!tag.requires_review? && !tag.usable? && 'batch-table__row--muted'] }
+  .batch-table__row__content.batch-table__row__content--padded.pending-account
+    .pending-account__header
```
I really don't like this relying on the pending-account UI, since it's just generic batch-table contents stuff by now, but I'm also not thinking that refactoring that would at all be in scope of this PR.

ThisIsMissEm commented on app/views/admin/tags/_tag.html.haml:
```diff
@@ -0,0 +1,28 @@
+    .pending-account__header
+      %strong
+        = link_to admin_tag_path(tag.id) do
+          = t('admin.tags.tag_name', name: tag.display_name)
```
Not sure using a translation here is correct, but it felt more correct than `= '#' + tag.display_name` or something like that.

ThisIsMissEm replied:
Interesting we have similar logic here: https://github.com/mastodon/mastodon/blob/74ead7d10698c7f18ca22c07d2f04ff81f419097/app/serializers/activitypub/hashtag_serializer.rb#L15

Perhaps it makes sense to have a `Tag#formatted_name` method?

oneiros replied:
I think it would!

And I would like that a lot better than the translation based approach.

ThisIsMissEm replied:
Okay, have implemented this, and also updated everywhere that we had this "##{...}" syntax for tags.

ThisIsMissEm commented on config/locales/en.yml:
```diff
@@ -880,7 +880,25 @@ en:
+      moderation:
+        not_trendable: Not trendable
+        not_usable: Not usable
+        pending_review: Pending review
+        review_requested: Review requested
```
There's a subtle difference between these two, arguably we could combine them, but I think we may want `Review requested` as a way to report tags as well in the future?

(Unless we change to polymorphic reports)

ThisIsMissEm commented on app/models/admin/tag_filter.rb:
```diff
@@ -0,0 +1,74 @@
+  end
+
+  def status_scope(value)
+    case value.to_s
```
@oneiros can the rails params return a value that's **not** a string?

oneiros replied:
Not in case of "regular" params sent by a browser. Since those are not typed, they will always be strings.

There is a special case, where a client can send JSON data and they end up in `params` as well. Since they are parsed as JSON, other data types may appear.

Of course this is not the case here. But there is another consideration: The controller could set a value before passing the params in here. I think here it already does to set a default order. One might want to allow symbols in this case, because they might look nicer in the controller. In that case `#to_s` would make sense here.

I would leave this up to you if you would like to allow this / guard against this here.

ThisIsMissEm commented on config/navigation.rb:
```diff
@@ -40,13 +40,14 @@
     n.item :trends, safe_join([fa_icon('fire fw'), t('admin.trends.title')]), admin_trends_statuses_path, if: -> { current_user.can?(:manage_taxonomies) && !self_destruct } do |s|
       s.item :statuses, safe_join([fa_icon('comments-o fw'), t('admin.trends.statuses.title')]), admin_trends_statuses_path, highlights_on: %r{/admin/trends/statuses}
-      s.item :tags, safe_join([fa_icon('hashtag fw'), t('admin.trends.tags.title')]), admin_trends_tags_path, highlights_on: %r{/admin/tags|/admin/trends/tags}
+      s.item :tags, safe_join([fa_icon('hashtag fw'), t('admin.trends.tags.title')]), admin_trends_tags_path, highlights_on: %r{/admin/trends/tags}
```
This change does mean that the `Trending > Hashtags` isn't highlighted when editing a tag from trending. The best we could do to get that back might be to use a query string or something?

```
/admin/tags/:id?trending=true
```

ClearlyClaire replied:
This seems like something we'll need to address later, but I'm not sure how at the moment.

ThisIsMissEm commented:
Okay, have updated this based on PR review feedback

oneiros commented on app/serializers/activitypub/actor_serializer.rb:
```diff
@@ -186,7 +186,7 @@ def href
 
     def name
-      "##{object.name}"
+      object.formatted_name
```
`Tag#formatted_name` uses the `#display_name` while this method used to use the `#name`. Those can be different (`name` being a normalized version of `display_name as per #18795).

So this here would actually change the behavior. (Of course I am in no way certain the old behavior was actually the intended one or an oversight in the linked PR :confused:).

ThisIsMissEm replied:
Hmm, good point! I'm not sure what the intended result should be here?

No tests began failing with this change 

cc @ClearlyClaire

ClearlyClaire replied:
That is a very good point. I suspect we might want `formatted_name` indeed, but I'd prefer that to be in another PR, as this does not seem related to the change at hand.

ClearlyClaire commented on app/models/admin/tag_filter.rb:
```diff
@@ -0,0 +1,74 @@
+    end
+  end
+
+  def status_scope(value)
+    case value.to_s
+    when 'reviewed'
+      Tag.reviewed
+    when 'review_requested'
+      Tag.pending_review
+    when 'unreviewed'
+      Tag.unreviewed
+    when 'trendable'
+      Tag.trendable
+    when 'not_trendable'
+      Tag.not_trendable
+    when 'usable'
+      Tag.usable
+    when 'not_usable'
+      Tag.not_usable
+    else
+      raise Mastodon::InvalidParameterError, "Unknown status: #{value}"
+    end
+  end
```
It seems to me many of these filters are orthogonal with each other and thus could be split. But this can be changed later if we need to, this does not need to block the PR.

ClearlyClaire commented on app/serializers/activitypub/note_serializer.rb:
This change should be in a different PR imho.

ClearlyClaire reviewed (changes requested):
This looks pretty good overall, I'd avoid the unrelated changes that I highlighted with inline comments, especially since some are more than refactoring and subtly change behavior, but otherwise this looks fine.

The filters could possibly be improved, but they seem useful already, and our moderators said they definitely see value in at least filtering out tags they already disabled.

ThisIsMissEm commented:
@ClearlyClaire I only did the formatted_name change following discussion with @oneiros — https://github.com/mastodon/mastodon/pull/30880#discussion_r1667188572

ClearlyClaire commented:
I can take over if you want.

ThisIsMissEm commented:
> I can take over if you want.

I'll see if I can get to this today, but I've also a bit of freelance work todo (fixing an outage/bug) + the activitypub & IETF 120 meetings.

@ClearlyClaire to be clear, we're saying revert most of the Tag#formatted_name changes?

ClearlyClaire commented:
> @ClearlyClaire to be clear, we're saying revert most of the Tag#formatted_name changes?

Yes. I'm fine with the refactor when there is no `name`/`display_name` change, what I want to avoid is the subtle behavior change. But I'm fine too with the `Tag#formatted_name` thing being left out of the PR completely. If you do leave it out, please use something like `"##{tag.display_name}"` rather than your initial i18n-based approach.

ThisIsMissEm committed:
Implement UI for Search & Moderation of Tags

ThisIsMissEm committed:
Add Tag#formatted_name for admin views

ThisIsMissEm commented:
@ClearlyClaire I've rebased this PR and split out the other Tag#formatted_name changes into #31170

ShadowJonathan commented:
Finally!! 🎉

nileshtrivedi commented:
@ThisIsMissEm Thanks for this PR!

I have a somewhat related question. On my instance, I am trying to leverage hashtags to simulate Slack's channels. These can be shared-across-instance, curated-by-admin fixed set of topics that can enable organization and discovery of old posts and conversations. Conceptually they are instance-level featured tags.

Is there a way to accomplish this for all client apps without modifying Mastodon server code? I have seen the code for the [Tag model](https://github.com/mastodon/mastodon/blob/main/app/models/tag.rb) and this does not seem possible. Mastodon has [account-specific featured hashtags](https://docs.joinmastodon.org/user/discoverability/#featured-tags), but because most users on my instance use [Elk client](https://github.com/elk-zone/elk), this probably would not work.

ThisIsMissEm commented:
@nileshtrivedi we currently don't have a notion of "recommended hashtags" though it has been suggested before I think, which would give you a page listing moderator recommended hashtags, which I think would accomplish what you want.

(it also makes sense following #31292) 

But hashtags still wouldn't be "instance specific" since all hashtags are global as far as I know

nileshtrivedi commented:
Thanks @ThisIsMissEm . I came across #21772 and submitted this feature request in #31505 .
